File: Array/zhanchengguo/0624-环形链表/Main.java
package com.test.ktapplication;

import android.util.Log;

class ListNode {
    int val;
    ListNode next;

    ListNode(int x) {
        val = x;
        next = null;
    }
}

class Solution {

    public void main() {
        ListNode a = new ListNode(1);
        ListNode b = new ListNode(1);
        ListNode c = new ListNode(1);
        ListNode d = new ListNode(1);
        a.next = b;
        b.next = c;
        c.next = d;
        d.next = null;
        boolean result = hasCycle(a);
        Log.w("result is ", result + "");
    }


    public boolean hasCycle(ListNode head) {
        if (head == null || head.next == null) {
            return false;
        }
        ListNode slow = head;
        ListNode fast = head.next;
        while (slow != fast) {
            if (fast == null || fast.next == null) {
                return false;
            }
            slow = slow.next;
            fast = fast.next.next;
        }
        return true;
    }
}

File: Array/zhanchengguo/0624-环形链表/main.kt
package com.test.ktapplication

class ListNode(var `val`: Int) {
    var next: ListNode? = null
}

fun main() {
    var a = ListNode(1)
    var b = ListNode(2)
    var c = ListNode(3)
    var d = ListNode(4)
    a.next = b
    b.next = c
    c.next = d
    d.next = c
    val result = hasCycle(a)
    print("result is ${result}")
}

fun hasCycle(head: ListNode?): Boolean {
    if (head?.next == null) {
        return false
    }
    var slow = head
    var fast = head.next
    while (slow != fast) {
        if (fast?.next == null) {
            return false
        }
        slow = slow?.next
        fast = fast?.next?.next
    }
    return true
}